packages/next-swc/crates/next-core/src/app_structure.rs

```diff
@@ -960,7 +977,20 @@ async fn directory_tree_to_entrypoints_internal_untraced(
     let subdirectories = &directory_tree.subdirectories;
     let components = directory_tree.components.await?.clone_value();
 
-    // if let Some(_) = components.page.or(components.default) {
+    // Route can have its own segment config, also can inherit from the layout root
+    // segment config. https://nextjs.org/docs/app/building-your-application/rendering/edge-and-nodejs-runtimes#segment-runtime-option
+    // if current tree have a layout segment config, treat it as _root_ segment
+    // config for the current route.
+    let root_segment_config = if let Some(layout) = components.layout {
+        let mut config = (*root_segment_config.await?).clone();
+        let source = Vc::upcast(FileSource::new(layout));
+        let layout_config = parse_segment_config_from_source(source);
```

Comment:
We should not parse the segment config in app_structure. This would require to parse all segment configs of the whole app.
Instead the config need to be parsed somewhere in write_to_disk so it's only parsed when the route is accessed.